File: p4.c
#include <stdio.h>

int ispal(int i)
{
    int dig[50], j = 0, k;

    for(; i > 0; i = i/10) {
        dig[j] = i%10;
        j += 1;
    }

    for(k = 0; k <= (j-1)/2; k++) {
        if(dig[k] != dig[j-1-k])
            return 0;
    }

    return 1;
}

int has3digfactors(int c)
{
    int i;
    for(i = 999; i > 1; i -= 1)
        if(c%i == 0 && c/i < 1000)
            return i;

    return 0;
}

int main(void)
{
    int c, f;

    for(c = 999*999; c > 0; c-=1)
        if(ispal(c) && (f = has3digfactors(c))) {
            printf("winner, winner: %d, factor = %d\n", c, f);
            break;
        }
}

File: go/p4.go
package main

import "fmt"

func isPal(n int) bool {
    s := fmt.Sprintf("%d", n)
    runes := []rune(s)
    rev := make([]rune, len(runes))
    copy(rev, runes)

    for i, j := 0, len(rev) - 1; i < j; i, j = i+1, j-1 {
        rev[i], rev[j] = rev[j], rev[i]
    }
    return string(runes) == string(rev)

}

func has3DigitFactors(n int) (bool, int) {
    for i := 999; i > 100; i-- {
        if n % i == 0 && n/i < 1000 && n/i > 99 {
            return true, i
        }

    }

    return false, 0

}

func main() {
    for c := 999*999; c > 9900; c-- {
        if isPal(c) {
            h3DF, factor := has3DigitFactors(c)
            if h3DF {
                fmt.Printf("winner, winner: %d\n", factor)
                break
            }
        }
    }

}
